pkg/algorithm/sku_scheduler.go

```diff
@@ -0,0 +1,517 @@
+// MIT License
+//
+// Copyright (c) Microsoft Corporation. All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE
+
+package algorithm
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/microsoft/hivedscheduler/pkg/api"
+	apiv2 "github.com/microsoft/hivedscheduler/pkg/api/v2"
+)
+
+// skuCell type for selected level cell in virtual cluster view
+type skuCell struct {
+	cell                          Cell            // within cell level, maybe higher or lower than node level
+	freeLeafCellNumAtPriority     int32           // free leaf cell number at the priority of the pod to be scheduled (lower priority considered as free)
+	usedLeafCellNumAtPriority     int32           // used leaf cell number at the priority of the pod to be scheduler
+	usedLeafCellNumHigherPriority int32           // used leaf cell number by higher priorities than the pod to be scheduler
+	healthy                       bool            // if the cell is healthy
+	address                       api.CellAddress // used for logging the cell address when bad or not suggested
+}
+
+// virtual cluster view
+type skuClusterView []*skuCell
+
+// skuScheduler can schedule a pod group of pods with arbitrary cell types on a cluster view.
```

Comment:
>with arbitrary cell types [](http://example.com/codeflow?start=49&length=25)

Also mentioned the withinOneCell feature?
Such as can be:
skuScheduler can schedule a group of pods with guaranteed affinity requirement (withinOneCell, e.g. with in one rack) and each pod can specify arbitrary cell types (e.g. non-leaf cell).

Reply:
updated